Feature/presentation layer eh

puch0handsup opened the pull request:
Added Data Layer.
-Dependency inyection modified.
-Modified Data type objects(Dto) obtained from API.
-Created Entity objects to communicate with RoomDB.

Added Domain Layer.
-Dependency inyection modified.
-Edited Meal Dto and Entity in order to map to Meal.
-Created the GetMealsByName.kt USE CASE, in order to retrieve meals.

Presentation Layer created.
-Created the MealSearchFragment.kt to display a list of meals by search.
-Created the MealsViewModel, that will communicate Presentation with Domain Layer.
-Added Navigation.

puch0handsup committed:
Data Layer created.
Dependency inyection added.
Data type objects(Dto) obtained from API.
Entity objects to communicate with RoomDB.

puch0handsup committed:
Domain Layer created.
Dependency inyection modified.
Edited Meal Dto and Entity in order to map to Meal.
Created the GetMealsByName.kt USE CASE, in order to retrieve meals.

puch0handsup committed:
Presentation Layer created.
Edited Meal Dto and Entity in order to map to Meal.
Created the MealSearchFragment.kt to display a list of meals by search.
Created the MealsViewModel, that will communicate Presentation with Domain Layer

inaki92 commented on app/src/main/java/com/example/themealsapp/data/remote/model/meal/MealDto.kt:
```diff
@@ -0,0 +1,165 @@
+    @SerializedName("strYoutube")
+    val strYoutube: String? = null
+){
+    fun toMealEntity(): MealEntity {
```
Use extension functions

puch0handsup replied:
I have actually tried to implement it, and it won't work. I will try again this evening. Thank you.

inaki92 replied:
resolved

inaki92 commented on app/src/main/java/com/example/themealsapp/data/local/entity/MealEntity.kt:
```diff
@@ -0,0 +1,106 @@
+    val strMeasure19: String?,
+    val strMeasure20: String?,
+){
+    fun toMeal(): Meal {
```
here use extension functions, the map should be inside this extension function

puch0handsup replied:
I have actually tried to implement it, and it won't work. I will try again this evening. Thank you.

inaki92 commented on app/src/main/java/com/example/themealsapp/data/remote/MealsAPI.kt:
```diff
@@ -1,16 +1,16 @@
-    ): Response<MealResponse>
+    @GET(SEARCH)
+    suspend fun searchMealByName(
+        @Query("s") query: String
```
lets give a meaningful name

puch0handsup replied:
Right!! Changed it.

inaki92 commented on app/src/main/java/com/example/themealsapp/data/repository/MealRepositoryImpl.kt:
```diff
@@ -0,0 +1,52 @@
+            if (response.isSuccessful) {
+                response.body()?.let { mealsResponse ->
+                    val mealInfos = mealsResponse.meals
+                    mealDao.deleteMeals(mealInfos?.map { it.strMeal })
```
No need if you have conflict strategy as REPLACE

puch0handsup replied:
You are right. Just Changed it

inaki92 commented on app/src/main/java/com/example/themealsapp/data/repository/MealRepositoryImpl.kt:
```diff
@@ -0,0 +1,52 @@
+                } ?: throw NullResponse()
+            } else throw FailureResponse(response.errorBody()?.string())
+        } catch (e: Exception) {
+            emit(UIState.ERROR(e))
```
here if there is an error, you can display the data base items, and tell user that you are using offline mode

puch0handsup replied:
Changed and put a flag to check if user has internet connection.

inaki92 commented on app/src/main/java/com/example/themealsapp/di/ApplicationModule.kt:
```diff
@@ -1,10 +1,67 @@
+        db: MealDatabase,
+        mealsAPI: MealsAPI
+    ): MealRepository {
+        return MealRepositoryImpl(mealDao = db.dao, mealsAPI = mealsAPI)
```
here you can provide the DAO directly

puch0handsup replied:
Figured that I could provide the DAO from the database

inaki92 commented on app/src/main/java/com/example/themealsapp/di/NetworkModule.kt:
```diff
@@ -1,37 +1,27 @@
         return Retrofit.Builder()
             .baseUrl(MealsAPI.BASE_URL)
-            .addConverterFactory(MoshiConverterFactory.create(moshi))
+            .addConverterFactory(GsonConverterFactory.create())
```
Why. did you decide to move to GSON? is there any advantage to using GSON? or disadvantage to using Moshi?

puch0handsup replied:
Even thougn Gson has a noticeable slower parsing time, they all about same for medium tests, but we do not know how big the database can be, in long response tests, surprisingly Moshi has the worst performance of the bunch.

d3v-w1n replied:
Good to know. I'll keep that in mind for future projects

inaki92 replied:
what is the speed parsing of gson vs moshi?

inaki92 commented on app/src/main/java/com/example/themealsapp/di/NetworkModule.kt:
```diff
@@ -49,15 +39,14 @@ class NetworkModule {
+    fun providesOkHttpLoggingInterceptor(): HttpLoggingInterceptor {
         return HttpLoggingInterceptor().apply {
-            level = HttpLoggingInterceptor.Level.BODY
+            level = HttpLoggingInterceptor.Level.BODY // Change it to NONE when app is launched to production
```
here you can use the debug flag from BUILDCONFIG to define the level

puch0handsup replied:
okkkk

inaki92 commented on app/src/main/java/com/example/themealsapp/domain/model/Meal.kt:
```diff
@@ -0,0 +1,51 @@
+    val strIngredient13: String,
+    val strIngredient14: String,
+    val strIngredient15: String,
+    val strIngredient16: String?,
```
if it is domain try not to have null values.  If it is needed try to have default values on nullables

puch0handsup replied:
I will try to change this in a next version, because I have already tried to work on it, and it did not work out.

inaki92 commented on app/src/main/java/com/example/themealsapp/presentation/view/MealSearchFragment.kt:
```diff
@@ -0,0 +1,71 @@
+
+        binding.rvMealSearch.apply {
+            layoutManager = LinearLayoutManager(
+                requireContext(),
```
why are you using require context instead of context or application context?

puch0handsup replied:
because I'm currently working on top of the activity. Pulling the application context would be too unnecessary.

inaki92 commented on app/src/main/java/com/example/themealsapp/presentation/view/MealSearchFragment.kt:
```diff
@@ -0,0 +1,71 @@
+        mealsViewModel.meals.observe(viewLifecycleOwner) { state ->
+            when (state){
+                is UIState.LOADING -> {
+
```
I guess there is a progress dialog you can use.

puch0handsup replied:
Progress Dialog is deprecated, but sure! I will ask @ArturoMarmolejo to add a ProgressBar in LOADING

inaki92 replied:
good

inaki92 commented on app/src/main/java/com/example/themealsapp/presentation/view/adapter/MealsListAdapter.kt:
```diff
@@ -0,0 +1,50 @@
+        binding.mealName.text = item.strMeal
+
+        itemView.setOnClickListener {
+            item.let { onItemClick }
```
looks clean

puch0handsup replied:
thank you?

inaki92 commented on app/src/main/java/com/example/themealsapp/presentation/viewmodel/MealsViewModel.kt:
```diff
@@ -0,0 +1,53 @@
+): ViewModel(){
+
+    init {
+        onSeach("")
```
lets check the method name, and the argument why is it empty string?

puch0handsup replied:
changed to "onSearchMealsByName()"

inaki92 commented on app/src/main/java/com/example/themealsapp/presentation/viewmodel/MealsViewModel.kt:
```diff
@@ -0,0 +1,53 @@
+
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+
+    fun onSeach(query: String) {
```
if empty string needed you can set the default value here

inaki92 replied:
query: String = ""

puch0handsup replied:
Yes, fixed! ty

inaki92 commented on app/src/main/java/com/example/themealsapp/utils/BaseFragment.kt:
```diff
@@ -0,0 +1,39 @@
+open class BaseFragment : Fragment() {
+
+    protected val mealsViewModel: MealsViewModel by lazy {
+        ViewModelProvider(requireActivity())[MealsViewModel::class.java]
```
whats the difference between require activity and activity variable?

puch0handsup replied:
It refers to the viewModelOwner. It will return the Activity where this fragment is associated with. It will not let you create the ViewModel if you pass Activity().

inaki92 commented on app/src/main/java/com/example/themealsapp/utils/NetworkState.kt:
```diff
@@ -0,0 +1,17 @@
+) {
+
+
+    fun isInternetOn() =
```
is this fully working?

d3v-w1n replied:
I will say I have seen the app function as intended with the project code in its current state - we should be good to go

puch0handsup replied:
it does

inaki92 commented on app/src/main/java/com/example/themealsapp/utils/UIState.kt:
```diff
@@ -1,7 +1,7 @@
 
 sealed class UIState<out T> {
-    object LOADING: UIState<Nothing>()
+    data class LOADING<T>(val message: T? = null): UIState<T>()
```
why you need the generic in the loading?

puch0handsup replied:
It is in order to be able to retrieve different objects using the same UIState. For instance, I am retrieving a List of Meals, in order to make the app scalable, I will need to retrieve List of Ingredients too.

inaki92 commented on app/src/main/res/layout/fragment_meal_search.xml:
```diff
@@ -0,0 +1,39 @@
+    android:layout_height="match_parent"
+    tools:context=".presentation.view.MealSearchFragment">
+
+    <!-- TODO: Update blank fragment layout -->
```
clean up your code

puch0handsup replied:
Sure, @ArturoMarmolejo  worked on it.

inaki92 commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,5 @@
 <resources>
     <string name="app_name">TheMealsApp</string>
+    <string name="search">Search</string>
+    <string name="hello_blank_fragment">Hello!</string>
```
clean up your code

puch0handsup replied:
Done

ArturoMarmolejo reviewed (approved):
Proceed with feedback comments and push again, if required. Confirming that Implementation and functionalities on this PR are currently working as expected. Approved without any further comments.

d3v-w1n commented on app/src/main/java/com/example/themealsapp/di/NetworkModule.kt:
```diff
@@ -1,37 +1,27 @@
 import okhttp3.logging.HttpLoggingInterceptor
 import retrofit2.Retrofit
-import retrofit2.converter.moshi.MoshiConverterFactory
+import retrofit2.converter.gson.GsonConverterFactory
```
This change in architecture was reviewed before lecture and maintains functionality of the Room DB implementation. After the code review, it has been discussed and the group chooses to keep it

d3v-w1n reviewed (approved):
The group now has the branches available for pull and will be able to rebase after merge approval - pending any changes that were requested during code review today

puch0handsup committed:
Errors for pull request fixed.

puch0handsup committed:
MealEntity.kt has been optimized in the function mapToMealEntity, so it doesn't store null values from neither ingredients, nor measurements. Instead, it will store a json in string format in the those fields. Also, instructions has been formatted to be split to a list.

Meal.kt has been modified when converting from ingredients or measures, it converts that string to a list. Furthermore, it also converts the instructions json string, to a list.

Changes have been made to improve general performance, and also provide more abstraction to the presentation layer.

inaki92 commented on app/src/main/java/com/example/themealsapp/data/local/entity/MealEntity.kt:
```diff
@@ -0,0 +1,92 @@
+        tempInstructions = tempInstructions?.replace("\r\n"," ")
+        val tempInstructions2 = tempInstructions?.split(". ")
+
+        val gson = Gson()
```
you have this guy in the module to be provided by hilt
